module/zfs/zio.c

```diff
@@ -4513,6 +4541,58 @@ zio_checksum_verify(zio_t *zio)
 	return (zio);
 }
 
+static zio_t *
+zio_dio_checksum_verify(zio_t *zio)
+{
+	zio_t *pio = zio_unique_parent(zio);
+
+	ASSERT3P(zio->io_vd, !=, NULL);
+	ASSERT3P(zio->io_bp, !=, NULL);
+	ASSERT3U(zio->io_child_type, ==, ZIO_CHILD_VDEV);
+	ASSERT3U(zio->io_type, ==, ZIO_TYPE_WRITE);
+	ASSERT3B(pio->io_prop.zp_direct_write, ==, B_TRUE);
+	ASSERT3U(pio->io_child_type, ==, ZIO_CHILD_LOGICAL);
+
+	if (zfs_vdev_direct_write_verify_pct == 0 || zio->io_error != 0)
+		goto out;
+
+	/*
+	 * A Direct I/O write checksum verification will only be
+	 * performed based on the top-level VDEV percentage for checks.
+	 */
+	uint32_t rand = random_in_range(100);
+	int error;
+
+	if (rand < zfs_vdev_direct_write_verify_pct) {
+		if ((error = zio_checksum_error(zio, NULL)) != 0) {
+			zio->io_error = error;
+			if (error == ECKSUM) {
+				mutex_enter(&zio->io_vd->vdev_stat_lock);
+				zio->io_vd->vdev_stat.vs_dio_verify_errors++;
+				mutex_exit(&zio->io_vd->vdev_stat_lock);
+				zio->io_error = SET_ERROR(EAGAIN);
+				zio->io_flags |= ZIO_FLAG_DIO_CHKSUM_ERR;
+
+				/*
+				 * The EAGAIN error must be propagated up to the
+				 * logical parent ZIO in zio_notify_parent() so
+				 * it can be returned to dmu_write_abd().
+				 */
+				zio->io_flags &= ~ZIO_FLAG_DONT_PROPAGATE;
+
+				(void) zfs_ereport_post(
+				    FM_EREPORT_ZFS_DIO_VERIFY,
+				    zio->io_spa, zio->io_vd, &zio->io_bookmark,
+				    zio, 0);
```

Comment:
So this concept came out of an OpenZFS Leadership Meeting. The idea here was that it was maybe too restrictive to keep this as a boolean module parameter. I personally don't care and am happy to convert this back to a boolean. In the world of HPC, our data is coming across a wire and there is no way for the data to be changed from a malicious user code at that point. So to me, this was a way to allow people who want to make sure there is no data corruption to be able to do that. The idea that came out of the Leadership Meeting is we could just do a percentage so a user could "possibly" detect that the checksum errors are tied to a malicious application.

I am happy to just turn this back to a boolean. I would prefer defaulting it to 0 though... However, I can see if we want to be strict about checking then okay, we can default to 1. I also am okay with just retuning `EIO` in this case. It might be cool to keep the counter even if returning `EIO` as it signals to the administrators there is a problem going on with an application if they set the parameter to 1. Again, this is not something I have to worry about in my line of work anyways.

@behlendorf what are your thoughts on this?

Reply:
I have commit ready to address this the way I suggested. Just using a boolean value and returning `EIO` on error. Ready to push it if @behlendorf agrees this is better idea and get feedback.